File: scripts/process.R
calc_pi <- function(N) {
    # Estimates pi using Monte-Carlo
    M <- 0
    for(i in 1:N) {
        x <- runif(1, -1, 1)
        y <- runif(1, -1, 1)

        if(x^2 + y^2 < 1) M <- M + 1
    }

    return(4 * M / N)
}


parser <- function(file) {
    # Parses the simulation settings from a file
    conn <- file(file, open="r")
    linn <-readLines(conn)
    close(conn)

    return(as.numeric(linn[1]))
}


writer <- function(val, file) {
    # Writes value to a file
    val_str <- format(val, scientific = FALSE)

    fileConn <- file(file)
    writeLines(val_str, fileConn)
    close(fileConn)
}


N <- parser(snakemake@input[['run']])
pi_approx <- calc_pi(N)
writer(pi_approx, snakemake@output[[1]])

File: scripts/process.py
import sys
import random

# ================== Functions ==================
def calc_pi(N):
    """ Estimates pi using Monte-Carlo """
    M = 0 # Initialize the counter
    for i in range(N):
        # Simulate impact coordinates
        x = random.uniform(-1, 1)
        y = random.uniform(-1, 1)

        # True if impact happens inside the circle
        if x**2 + y**2 < 1.0:
            M += 1

    return 4 * M / N


def parser(file):
    """ Parses the simulation settings from file """
    # Our file is remarkably simple: it just contains one integer
    #
    # The parser just extracts that number ...
    val = open(file, 'r').read()

    # ... and returns it as an integer
    return(int(val))


def writer(val, file):
    """ Writes a value to a file """
    val_str = f'{val:.8f}' # Keep decimals enough

    with open(file, 'w') as f:
        f.write(val_str)


N = parser(snakemake.input['run'])
pi_approx = calc_pi(N)
writer(pi_approx, snakemake.output[0])
